### PR TITLE
feat: register custom formats take 2

### LINKED ISSUE #5848: Custom output formatter
**Is your feature request related to a problem? Please describe.**

I'm setting up Netlify CMS for a mostly-static site, written in Next.js and TypeScript. None of the output types seem _quite_ right for the project:

- `md`: requires using a frontmatter parser, either at build time or runtime - and what's parsed will be typed as `any`
- `toml`: similar to above
- `yaml`: similar to above
- `json`: can be directly `import`-ed and will be strongly-typed, so this is the best option for right now. But multiline markdown blobs are hard to read, compared to `md` or `yaml`

And for _all_ of the above, any markdown has to be run through a markdown-to-jsx type library (we're currently using [markdown-to-jsx](https://npmjs.com/package/markdown-to-jsx)).

All of the above problems are surmountable, and most likely even without perf hits, by some combination of babel, webpack, and static-page-generation. But it feels like there's a trick being missed, because any processing needed of the content _could_ be done when it's generated, rather than leaving it to build-time or runtime, either server-side or client-side.

**Describe the solution you'd like**

The output that would probably be most useful for our team specifically would be something like a data-only typescript file:

```ts
import stripIndent from 'strip-indent'

export const en = {
  title: 'Contact Page',
  body: stripIndent(`
    ## Contact Us
    
    You can contact us at foo@bar.com. Visit our [help page](/help) for more info.
  `)
}

export const es = { ...spanish equivalent... } 
```

But of course, that's very project-specific. It depends on the `strip-indent` library, which is fine for our team, but isn't in general. So I'm not proposing netlify-cms include a typescript formatter. But it'd be great if you could let us define our own. I'm imagining an API something like:

```js
CMS.registerFormatter({
  extension: '.ts',
  formatter: {
    fromFile: content => {
       // whatever magic is required... we can use eval at our own risk, I guess
      let _exports = {}; 
      eval(content.replace(/export const (\w+) = /g, '_exports.$1 = '))
      return exports
    },
    toFile: data => {
      // construct a typescript ast somehow... again, we can make project-specific assumptions here
      return Object.entries(data)
        .map(([name, value]) => `export const ${name} = ${JSON.stringify(value, null, 2)}`)
        .join('\n\n')
    },
  },
})
```

This formatter would then be inferred for pages with `file: foobar.ts`.

The actual API of a formatter is really nice and simple (I'm basing this on the [json example](https://github.com/netlify/netlify-cms/blob/master/packages/netlify-cms-core/src/formats/json.ts)) so it's easy to imagine users being able to get pretty creative with how the output file gets persisted, to make sure it's a combination of human-readable and useful for the project it lives in.

If users wanted to, they could even do markdown processing in the `toFile` function to build up a ready-to-render AST.

**Describe alternatives you've considered**

Conceivably this could be done by [wrapping an existing backend](https://www.netlifycms.org/docs/backends-overview/#creating-a-new-backend) but that seems more invasive and scary. There might also be a way to do this already, I haven't found it in the docs though.

There might also be a hacky way to do `window.CMS.something.formatters.toFile = ...` but I'm not familiar enough with the codebase to know how/whether formatters are exposed.

**Additional context**

I'm new to Netlify CMS, so please let me know if I'm thinking about things in The Wrong Way.

### PR DESCRIPTION
**Pull request by mmkal**
Replaces #5839
Fixes #5848

**Summary**

Adds the ability to register custom formats (could be xml, po, binary, json5, etc.).

Also adds the requested tests and docs from the last PR (1dc43db87936fa480d3650972f2bf0a0882da726 + 6d29b239cb36073d4dde8d5bd0d3db87c86eab07)

**Test plan**

Unit tests for register function + formatters themselves. Then ran manual tests by doing `npm pack`, installed the `.tgz` file directly in the project we're using netlify-cms with, then registering some custom formats... and it worked.

**Checklist**

- [x] I have read the [contribution guidelines](../CONTRIBUTING.md).
- [x] Code is formatted via running `yarn format`.
- [x] Tests are passing via running `yarn test`.
- [x] The status checks are successful (continuous integration). Those can be seen below.

### CODE REVIEW

**Comment by mmkal** [comment]
@erezrokah any ideas why cypress is failing? I wasn't able to run them myself and don't have permission to download screenshots.

**Comment by erezrokah** [comment]
Looks like our tests started failing with a specific renovate PR https://github.com/netlify/netlify-cms/actions/workflows/nodejs.yml?query=branch%3Amaster.

Looking into it

**Comment by mmkal** [comment]
@erezrokah using `yarn test:e2e:dev` the `can change status on and publish multiple entries` test was passing for me locally, so I don't think the failure is related to these changes, unless I'm misunderstanding something?

**Comment by erezrokah** [comment]
> @erezrokah using `yarn test:e2e:dev` the `can change status on and publish multiple entries` test was passing for me locally, so I don't think the failure is related to these changes, unless I'm misunderstanding something?

You're correct. I don't think the failure is related to your changes. However I'd prefer to fix the tests before merging this PR.

**Comment by erezrokah** [comment]
> It says there's one workflow awaiting approval

That's GitHub's crypto mining defense technic :) First committers need approval from a maintainer to run the CI.

I'm trying to think of a way to avoid manual initialization. We could have the custom formats defined as top level item `config.yml` and use that to validate, WDYT? 

It will also make it clear what's supported

**Comment by mmkal** [comment]
> > It says there's one workflow awaiting approval
> 
> That's GitHub's crypto mining defense technic :) First committers need approval from a maintainer to run the CI.
> 
> I'm trying to think of a way to avoid manual initialization. We could have the custom formats defined as top level item `config.yml` and use that to validate, WDYT?
> 
> It will also make it clear what's supported

I'm not sure - wouldn't some custom javascript have to run to register the formatter implementation? Where would netlify-cms know where to find that javascript if there's no manual initialization. Also - I don't feel very comfortable adding non-manual initialization support since the team have been using manual initialization since day one.

Would you be open to it going in without, then if there are a users requesting custom formats without wanting to manually initialize, a separate piece of work could be done to support that use case too? (Selfishly - that wouldn't help my team and we would _love_ to be able to use this now!)

**Comment by erezrokah** [comment]
Thanks for your input @mmkal, what concerns me is that the CMS default is not to have manual initialization so requiring it here breaks that. Also other `register*` methods don't require it.

Another option is to drop the static config level schema validation and only validate dynamically when trying to retrieve the format. We do the latter for custom widgets and display an "Unknown" widget when that happens.

WDYT?

**Comment by mmkal** [comment]
> Thanks for your input @mmkal, what concerns me is that the CMS default is not to have manual initialization so requiring it here breaks that. Also other `register*` methods don't require it.

I'm not sure I agree it breaks that. The default setup still doesn't require manual initialisation - it's only required if you want to start pretty seriously customising the behaviour by adding extra file formats. Having said that, I can't see why it needs to be any different from registering a custom widget. 

> Another option is to drop the static config level schema validation and only validate dynamically when trying to retrieve the format. We do the latter for custom widgets and display an "Unknown" widget when that happens.
> 
> WDYT?

I think this is already happening thanks to this change:

image

If someone tries to use an un-registered widget, there will just be an error thrown. So maybe this is already meeting the requirements? Is there a simple way to test this with a non-manual-initialized setup?

**Comment by mmkal** [comment]
@erezrokah just tested this and it actually doesn't require manual initialization. It works similarly to the other `CMS.register...` methods.

It looks like this if you try to use a non-existent formatter:

image

So to address your two comments:

1. Manual initialization: not required.
2. Drop static validation: done. We now only validate that the prop is a string.

So, anything else needed for this to go in?? 🙏

**Comment by mmkal** [comment]
@erezrokah bump - anything I can do to get this in?

**Comment by mmkal** [comment]
Ping @erezrokah - CC @bytrangle in case Erez is no longer working on netlify-cms?

**Comment by erezrokah** [comment]
pinging @krider2010

**Comment by bytrangle** [comment]
@mmkal Thank you for the time you've put into this PR. I will find time to review this soon.

**Review by bytrangle — COMMENTED** [review]
Very well-thought out.This seems like a textbook example for how to make a PR.

File: packages/netlify-cms-core/src/reducers/collections.ts
Comment (mmkal) [reply packages/netlify-cms-core/src/reducers/collections.ts]: Switch to Error [2f95d52](https://github.com/netlify/netlify-cms/pull/6205/commits/2f95d52cec5900572918ec801cbf7da8a769dd89)

**Comment by mmkal** [comment]
@bytrangle @krider2010 @erezrokah fixed the prettier problem - hopefully ready now but won't know til the build is approved! 🙏

**Comment by mmkal** [comment]
I still want to merge this!

**Comment by mmkal** [comment]
@martinjagodic the only thing failing (after a year and a half of "Update from main and cross fingers") is `Markdown widget breaks pressing enter`:

image
 
I really don't think it's related - could we merge this as is? I've updated to incorporate all the `netlify-cms-*` to `decap-cms-*` package changes.

**Review by martinjagodic — APPROVED** [review]

**Pull request merged**
